apps/prairielearn/src/lib/groups.ts

```diff
@@ -229,21 +229,28 @@ export async function addUserToGroup({
     }
 
     const user = await selectUserByUid(uid);
-    const enrollment = user
-      ? await getEnrollmentForUserInCourseInstance({
+    // To be part of a group, the user needs to either be enrolled in the course
+    // instance, or be an instructor
+    const isInCourseInstance = user
+      ? (await getEnrollmentForUserInCourseInstance({
           course_instance_id: group.course_instance_id,
           user_id: user.user_id,
-        })
-      : null;
-    if (enrollment == null) {
+        })) != null ||
+        (await sqldb.callRow(
+          'users_is_instructor_in_course_instance',
+          [user.user_id, group.course_instance_id],
+          z.boolean(),
+        ))
+      : false;
```

Comment:
Nit: I find this pretty hard to read. I'd say either a) lift this into a function where you can take advantage of early returns, or b) use use normal `if ...` to handle the branches.

Reply:
I tried with normal `if` statements, but that made TS unable to identify that `user` is not null after this statement due to the multiple conditions. I have split the tests in 7c5492cd45869c452f54410a1ed6f69ee23b93c3, still taking advantage of short-circuit, but testing for instructor first.